worlds/pokemon_crystal/rom.py

```diff
@@ -0,0 +1,425 @@
+
+
+class PokemonCrystalProcedurePatch(APProcedurePatch, APTokenMixin):
+    game = "Pokemon Crystal"
+    hash = "9f2922b235a5eeb78d65594e82ef5dde"
+    patch_file_ending = ".apcrystal"
+    result_file_ending = ".gbc"
+
+    procedure = [
+        ("apply_bsdiff4", ["basepatch.bsdiff4"]),
+        ("apply_tokens", ["token_data.bin"])
+    ]
+
+    @classmethod
+    def get_source_data(cls) -> bytes:
+        return get_base_rom_as_bytes()
+
+
+def generate_output(world: "PokemonCrystalWorld", output_directory: str, patch: PokemonCrystalProcedurePatch) -> None:
+    item_texts = []
+    for location in world.multiworld.get_locations(world.player):
+        if location.address is None:
+            continue
+
+        if location.item and location.item.player == world.player:
+            item_id = location.item.code
+            item_id = item_id - 256 if item_id > 256 else item_id
+            write_bytes(patch, [item_id], location.rom_address)
+        else:
+            # for in game text
+            item_flag = location.address
+            player_name = world.multiworld.player_name[location.item.player].upper()
+            item_name = location.item.name.upper()
+            item_texts.append([player_name, item_name, item_flag])
+
+            write_bytes(patch, [item_const_name_to_id("AP_ITEM")], location.rom_address)
+
+    # table has format: location id (2 bytes), string address (2 bytes), string bank (1 byte),
+    # and is terminated by 0xFF
+    item_name_table_length = len(item_texts) * 5 + 1
+    item_name_table_adr = data.rom_addresses["AP_ItemText_Table"]
+
+    # strings are 16 chars each, plus a terminator byte,
+    # this gives every pair of item + player names a size of 34 bytes
+    item_name_bank1 = item_name_table_adr + item_name_table_length
+    item_name_bank1_length = data.rom_addresses["AP_ItemText_Bank1_End"] - item_name_bank1
+    item_name_bank1_capacity = int(item_name_bank1_length / 34)
+
+    item_name_bank2 = data.rom_addresses["AP_ItemText_Bank2"]
+    item_name_bank2_length = data.rom_addresses["AP_ItemText_Bank2_End"] - item_name_bank2
+    item_name_bank2_capacity = int(item_name_bank2_length / 34)
+
+    for i, text in enumerate(item_texts):
+        # truncate if too long
+        player_text = convert_to_ingame_text(text[0])[:16]
+        # pad with terminator byte to keep alignment
+        player_text += [0x50] * (17 - len(player_text))
+        item_text = convert_to_ingame_text(text[1])[:16]
+        item_text.append(0x50)
+        # bank 1
+        bank = 0x75
+        table_offset_adr = item_name_table_adr + i * 5
+
+        if i >= item_name_bank1_capacity + item_name_bank2_capacity:
+            # if we somehow run out of capacity in both banks, just finish the table and break,
+            # there is a fallback string in the ROM, so it should handle this gracefully.
+            write_bytes(patch, [0xFF], item_name_table_adr + table_offset_adr)
+            print("oopsie")
```

Comment:
Lol, if leaving this in was intentional, it should probably be a `logging.warning` or `debug` or something with a clearer message.